### PR TITLE
[ML][Transforms] adjusting stats.progress for cont. transforms

### LINKED ISSUE #44557: [ML] Data frame progress is not reporting for continuous
Found in 7.3.0-SNAPSHOT `{
  "build" : {
    "hash" : "a57a5c5",
    "date" : "2019-07-16T14:52:05.956252Z"
  },`

Checkpoint progress is not being updated for checkpoints 1 and above, presumably since the introduction of page-by-page processing to avoid terms explosion. Progress is reported and updated for checkpoint 0.

This is related to https://github.com/elastic/elasticsearch/issues/43767. We may wish to revisit how we can show indicative checkpoint progress for continuous data frames. For example, 20% progress at page 2 of 10 could be a good enough indicator.

In debug logging you can see the following

```
[2019-07-18T11:03:33,859][DEBUG][o.e.x.c.i.AsyncTwoPhaseIndexer] [node1] Beginning to index [df01], state: [STARTED]
[2019-07-18T11:03:34,944][DEBUG][o.e.x.d.c.DataFrameTransformsCheckpointService] [node1] Failed to retrieve source checkpoint for data frame [df01]
[2019-07-18T11:03:35,556][DEBUG][o.e.x.d.t.DataFrameTransformTask] [node1] Updating persistent state of transform [df01] to [{"task_state":"started","indexer_state":"indexing","position":{"indexer_position":{"clientip":"2.124.127.204"},"bucket_position":{"clientip":"2.122.228.0"}},"checkpoint":5,"progress":{"total_docs":32000,"docs_remaining":15692,"percent_complete":50.9625}}]
[2019-07-18T11:03:40,741][DEBUG][o.e.x.d.t.DataFrameTransformTask] [node1] Updating persistent state of transform [df01] to [{"task_state":"started","indexer_state":"indexing","position":{"indexer_position":{"clientip":"31.51.217.58"},"bucket_position":{"clientip":"31.49.140.121"}},"checkpoint":5,"progress":{"total_docs":32000,"docs_remaining":0,"percent_complete":100.0}}]
[2019-07-18T11:03:46,353][DEBUG][o.e.x.d.t.DataFrameTransformTask] [node1] Updating persistent state of transform [df01] to [{"task_state":"started","indexer_state":"indexing","position":{"indexer_position":{"clientip":"49.212.153.96"},"bucket_position":{"clientip":"46.252.71.2"}},"checkpoint":5,"progress":{"total_docs":32000,"docs_remaining":0,"percent_complete":100.0}}]
[2019-07-18T11:03:51,018][DEBUG][o.e.x.d.t.DataFrameTransformTask] [node1] Updating persistent state of transform [df01] to [{"task_state":"started","indexer_state":"indexing","position":{"indexer_position":{"clientip":"64.246.178.34"},"bucket_position":{"clientip":"64.246.161.190"}},"checkpoint":5,"progress":{"total_docs":32000,"docs_remaining":0,"percent_complete":100.0}}]
[2019-07-18T11:03:55,571][DEBUG][o.e.x.d.t.DataFrameTransformTask] [node1] Updating persistent state of transform [df01] to [{"task_state":"started","indexer_state":"indexing","position":{"indexer_position":{"clientip":"70.63.164.253"},"bucket_position":{"clientip":"70.29.234.38"}},"checkpoint":5,"progress":{"total_docs":32000,"docs_remaining":0,"percent_complete":100.0}}]
[2019-07-18T11:03:59,916][DEBUG][o.e.x.d.t.DataFrameTransformTask] [node1] Updating persistent state of transform [df01] to [{"task_state":"started","indexer_state":"indexing","position":{"indexer_position":{"clientip":"80.128.104.91"},"bucket_position":{"clientip":"80.87.25.79"}},"checkpoint":5,"progress":{"total_docs":32000,"docs_remaining":0,"percent_complete":100.0}}]
[2019-07-18T11:04:03,630][DEBUG][o.e.x.d.t.DataFrameTransformTask] [node1] Updating persistent state of transform [df01] to [{"task_state":"started","indexer_state":"indexing","position":{"indexer_position":{"clientip":"82.145.209.89"},"bucket_position":{"clientip":"82.132.245.233"}},"checkpoint":5,"progress":{"total_docs":32000,"docs_remaining":0,"percent_complete":100.0}}]
[2019-07-18T11:04:06,957][DEBUG][o.e.x.d.t.DataFrameTransformTask] [node1] Updating persistent state of transform [df01] to [{"task_state":"started","indexer_state":"indexing","position":{"indexer_position":{"clientip":"86.161.217.16"},"bucket_position":{"clientip":"86.161.54.214"}},"checkpoint":5,"progress":{"total_docs":32000,"docs_remaining":0,"percent_complete":100.0}}]
[2019-07-18T11:04:10,428][DEBUG][o.e.x.d.t.DataFrameTransformTask] [node1] Updating persistent state of transform [df01] to [{"task_state":"started","indexer_state":"indexing","position":{"indexer_position":{"clientip":"91.192.111.141"},"bucket_position":{"clientip":"91.125.129.31"}},"checkpoint":5,"progress":{"total_docs":32000,"docs_remaining":0,"percent_complete":100.0}}]
[2019-07-18T11:04:14,504][DEBUG][o.e.x.d.t.DataFrameTransformTask] [node1] Updating persistent state of transform [df01] to [{"task_state":"started","indexer_state":"indexing","position":{"indexer_position":{"clientip":"101.226.168.240"},"bucket_position":{"clientip":"101.226.166.216"}},"checkpoint":5,"progress":{"total_docs":32000,"docs_remaining":0,"percent_complete":100.0}}]
[2019-07-18T11:04:17,897][DEBUG][o.e.x.d.t.DataFrameTransformTask] [node1] Updating persistent state of transform [df01] to [{"task_state":"started","indexer_state":"indexing","position":{"indexer_position":{"clientip":"157.55.34.107"},"bucket_position":{"clientip":"157.55.33.249"}},"checkpoint":5,"progress":{"total_docs":32000,"docs_remaining":0,"percent_complete":100.0}}]
[2019-07-18T11:04:21,134][DEBUG][o.e.x.d.t.DataFrameTransformTask] [node1] Updating persistent state of transform [df01] to [{"task_state":"started","indexer_state":"indexing","position":{"indexer_position":{"clientip":"188.29.164.205"},"bucket_position":{"clientip":"188.29.37.197"}},"checkpoint":5,"progress":{"total_docs":32000,"docs_remaining":0,"percent_complete":100.0}}]
[2019-07-18T11:04:23,961][DEBUG][o.e.x.c.i.AsyncTwoPhaseIndexer] [node1] Finished indexing for job [df01], saving state and shutting down.
[2019-07-18T11:04:23,961][DEBUG][o.e.x.d.t.DataFrameTransformTask] [node1] Finished indexing for data frame transform [df01] checkpoint [5]
```

### PR DESCRIPTION
**Pull request by benwtrent**
The idea of "number of documents to process" does not make sense for continuous transforms. Every checkpoint after the initial batch process can be short lived. To provide a better sense of "work being done" and "how much work is left", some more statistics are being gathered and returned. 

The following fields are now optional in checkpoint_progress, depending on if the checkpoint is the first batch one or not:
* percent_complete
* docs_remaining
* total_docs

The following fields are added to checkpoint_progress
* documents_processed
* documents_indexed

The following has been added to the stats object
* exponential_avg_documents_indexed
* exponential_avg_documents_processed
* exponential_avg_checkpoint_duration_ms
* continuous_checkpoints_processed

The following has been added to the checkpointing object
* found_changes
* last_change_check

These new statistics should give a better, fuller view of the current checkpoints work, the previous work completed, and an expectation of how long a given checkpoint should take. 

closes https://github.com/elastic/elasticsearch/issues/44557

### CODE REVIEW

**Comment by benwtrent** [comment]
@walterra ping, API changes

**Comment by benwtrent** [comment]
I need to verify we have facility for updating mappings. A continuous transform that moves from a 7.3 node to a 7.4 node (once this is backported), will need to potentially index new values into the transforms index...

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/dataframe/transforms/DataFrameIndexerTransformStats.java [hunk 1]
```diff
@@ -102,23 +147,49 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
...
+        assert checkpointDurationMs >= 0;
+        assert docsIndexed >= 0;
+        assert docsProcessed >= 0;
+        continuousCheckpointsProcessed++;
+        if (continuousCheckpointsProcessed == 1) {
+            expAvgCheckpointDurationMs = checkpointDurationMs;
+            expAvgDocumentsIndexed = docsIndexed;
+            expAvgDocumentsProcessed = docsProcessed;
+        } else {
+            double alpha = 2.0/(continuousCheckpointsProcessed+1);
```
Comment (davidkyle) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/dataframe/transforms/DataFrameIndexerTransformStats.java]: The more checkpoints processed the smaller `alpha` gets and the less significant the latest value becomes.  Did you consider using a constant for `alpha`?

**Review by davidkyle — COMMENTED** [review]
Code changes LGTM. 

This adds new fields to the already busy `stats` documents I just wanted to be clear about what they represent.

**found_changes** and **last_change_check**
This tells me whether or not the index changed in the last 60 seconds or whatever the sync period is. I worry that any state changes will be brief (mostly `found_changes` is false but every N buckets it flips to true for single bucket) maybe `last_time_a_change_was_found` is more informative.  


> The following fields are added to checkpoint_progress
- documents_processed
- documents_indexed

I'm not sure how much value `documents_indexed` adds if `documents_processed` is being use to show the transform is doing work

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/dataframe/transforms/DataFrameIndexerTransformStats.java (hunk 1 above)
Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/dataframe/transforms/DataFrameIndexerTransformStats.java]: @davidkyle you are absolutely correct. I mixed up my implementations. I was initially doing some fancy windowing, but determined it was no longer necessary and then did not change my calculation to reflect that. 

Will update it for sure.

**Comment by benwtrent** [comment]
@davidkyle 

`last_time_a_change_was_found` <- Love it, I will use something like this instead.  Much better.

> I'm not sure how much value `documents_indexed` adds if `documents_processed` is being use to show the transform is doing work

The run-time of a transform checkpoint is dictated by two things: Search time + indexing time; `documents_processed` and `documents_indexed` should be highly correlated to each (respectively). Having both values gives the user insight into "how much work" (i.e. how many documents are being read and then indexed) the transform is doing per checkpoint and how long that work actually takes.

**Comment by davidkyle** [comment]
@elasticmachine update branch

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/dataframe/transforms/DataFrameTransformCheckpointingInfo.java [hunk 2]
```diff
@@ -65,18 +79,31 @@
      * @param last stats of the last checkpoint
      * @param next stats of the next checkpoint
      * @param operationsBehind counter of operations the current checkpoint is behind source
+     * @param changesLastDetectedAt the last time the source indices were checked for changes
```
Comment (davidkyle) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/dataframe/transforms/DataFrameTransformCheckpointingInfo.java]: nit

```suggestion
     * @param changesLastDetectedAt the last time changes were detected in the source indices
```

**Review by davidkyle — APPROVED** [review]
LGTM

You may want a second opinion however

**Pull request merged**